cmake/ccache.cmake

```diff
@@ -0,0 +1,93 @@
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+]]
+
+include_guard(GLOBAL)
+
+
+#[=======================================================================[.rst:
+enable_ccacche
+------------------
+
+Overview
+^^^^^^^^
+
+Enable use of Ccache for compiler caching during the build process.
+
+.. code-block:: cmake
+
+  enable_ccacche(
+      [QUIET]
+  )
+   -- Searches for the ccache executable, and if found enables it for supported
+   language with specified options.
+
+  ``QUIET`` Silent output from this method.
+
+#]=======================================================================]
+function(enable_ccache)
+    set(options QUIET)
+    set(oneValueArgs)
+    set(multiValueArgs)
+    cmake_parse_arguments(CCACHE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    find_program(CCACHE_BIN ccache REQUIRED)
+    if(NOT CCACHE_BIN)
+        message(STATUS "Morpheus: CCache not found. Rebuilding all the source files.")
+        message(DEBUG "Morpheus: If CCache compiler cache is desired then ensure this is a supported platform and ensure the Conan buildenv is active")
+        return()
+    endif()
+
+    message(STATUS "Morpheus Generator: ${CMAKE_GENERATOR}") #debug
+    message(STATUS "Morpheus: CCache found: ${CCACHE_BIN}. Enabling ccache as compiler cache tool.")
+    if(${CMAKE_GENERATOR} MATCHES "Ninja|Makefiles")
+        message(STATUS "found generator Ninja|Makefiles") #debug
+        foreach(lang IN ITEMS ${MORPHEUS_LANGUAGES})
+            set(CMAKE_${lang}_COMPILER_LAUNCHER ${CCACHE_BIN})
+        endforeach()
+    elseif(${CMAKE_GENERATOR} STREQUAL "Xcode")
+        message(STATUS "found generator Xcode") #debug
+        foreach(lang IN ITEMS ${MORPHEUS_LANGUAGES})
+            set(launch${lang} ${CMAKE_BINARY_DIR}/launch-${lang})
+            file(WRITE ${launch${lang}}
+                "#!/bin/bash\n\n"
+                "exec \"${CCACHE_BIN}\" \"${CMAKE_${lang}_COMPILER}\" \"$@\"\n"
+            )
+            execute_process(COMMAND chmod a+rx ${launch${lang}})
+            set(${CMAKE_XCODE_ATTRIBUTE_${lang}} ${launch${lang}})
+            if(lang STREQUAL "C")
+                set(CMAKE_XCODE_ATTRIBUTE_LD ${launchC}) 
+            elseif(lang STREQUAL "CXX")
+                set(CMAKE_XCODE_ATTRIBUTE_LDPLUSPLUS ${launchCXX})
+            endif()
+        endforeach()
+    elseif(${CMAKE_GENERATOR} MATCHES "Visual Studio")
```

Comment:
I imagine with Visual Studio support we have to limit the languages to language implicitly supported by the MSbuild engine?

Reply:
Ah probably yes, sorry but I am missing this one. MSBuild engine supports C and CXX I suppose. The documentation is [there](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild?view=vs-2022) and from a quick sight It seems supportin C with MSBuild and CXX with MSBuild++.

Reply:
Yes, and Cuda should be good. But I'd expect Objective C/C++ to not work here.  Main point here is I think we need to explicit test support for each language (and be we I mean me as I have a Windows Laptop with Visual Studio installed which I can test on)

Reply:
Hey, that's pretty close (Heisenberg used Conan 1).  I should write up a page for Windows instructions too.  Let me know if you hit any issues trying this out.